🚀 2단계 - 수강신청(도메인 모델)

yoohojeong opened the pull request:
🧑‍💻수강 신청 기능 요구사항
---
**CoverImage**
> - [X] 이미지 크기는 1MB 이하여야 한다.
> - [X] 이미지 타입은 gif, jpg(jpeg 포함),, png, svg만 허용한다.
> - [X] 이미지의 width는 300픽셀, height는 200픽셀 이상이어야 하며, width와 height의 비율은 3:2여야 한다.

**Session**
> - [X] 강의는 무료 강의와 유료 강의로 나뉜다.
> - [X] 강의는 강의 커버 이미지 정보를 가진다.
> - [X] 무료 강의는 최대 수강 인원 제한이 없다.
> - [X] 강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.

**Period**
> - [X] 강의는 시작일과 종료일을 가진다.

**SessionFee** (추가)
> - [X] 강의는 무료 강의와 유료 강의로 나뉜다.

**Enrollment**
> - [X] 유료 강의의 경우 결제는 이미 완료한 것으로 가정하고 이후 과정을 구현한다.
> - [X] 강의 수강신청은 강의 상태가 모집중일 때만 가능하다.
> - [X] 유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다.

**SessionUser**
> - [X] 수강신청 완료 시 Session, NsUser 정보를 등록한다.

**SessionUsers**
> - [X] 유료 강의는 강의 최대 수강 인원을 초과했을 경우 더이상 수강신청 할 수 없다.

**Course**
**Sessions**
> - [X] 과정(Course)은 기수 단위로 운영하며, 여러 개의 강의(Session)를 가질 수 있다.

**Payment**
> - [X] 결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.

🔥 테스트 결과
---
> <img width="472" alt="스크린샷 2023-12-10 오후 9 36 35" src="https://github.com/next-step/java-lms/assets/23239084/1652aaf7-c3ca-400b-a190-15d6ae0d3f00">

yoohojeong committed:
feat : SessionImage 도메인 구현 / Sisson 내 시작일, 종료일 추가

yoohojeong committed:
feat : Session 도메인 구현
- 강의는 시작일과 종료일을 가진다.
- 강의는 강의 커버 이미지 정보를 가진다.
- 강의는 무료 강의와 유료 강의로 나뉜다.
- 무료 강의는 최대 수강 인원 제한이 없다.
- 강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.

yoohojeong committed:
feat : Period 도메인 구현

yoohojeong committed:
feat : Enrollment 도메인 구현 완료

yoohojeong committed:
feat : SessionUser, SessionUsers 도메인 구현

yoohojeong committed:
fix : NPE 해결

yoohojeong committed:
feat : EnrollmentService 틀 구현

yoohojeong committed:
style : Readme 변경

yoohojeong committed:
feat : #248 피드백 반영
- Answers에서 Answer 객체에게 없는 필요한 데이터를 주입해주면서 히스토리 생성을 명령

ohtaeg commented on src/main/java/nextstep/courses/domain/sessionuser/SessionUsers.java:
```diff
@@ -0,0 +1,24 @@
+
+public class SessionUsers {
+
+    private List<SessionUser> sessionUsers = new ArrayList<>();
```
같은 학생이 중복으로 등록이 가능한 구조인 것 같아요..!

yoohojeong replied:
중복 생성을 막기 위해 List에서 Set으로 타입 변경했습니다~!

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,80 @@
+    private final String id;
+    private final FreeOrPaid freeOrPaid;
+    private final CoverImage coverImage;
+    private final Period period;
```
시작일과 종료일을 하나의 객체로 다뤄주셨네요 👍 👍

ohtaeg commented on src/main/java/nextstep/courses/domain/course/Course.java:
```diff
@@ -0,0 +1,55 @@
+    private Sessions sessions;
+
+    public Course(String title, Long creatorId) {
+        this(0L, title, creatorId, LocalDateTime.now(), null, null);
```
title, creatorId가 다른 과정을 만들어도 id는 같을 수 있겠군요 👀

yoohojeong replied:
오 넵 아래와 같이 id 자리에 0L이 아닌 null을 대입해서 auto generated 되게 하였습니다!
`this(null, title, creatorId, LocalDateTime.now(), null, null);`

<img width="621" alt="스크린샷 2023-12-11 오전 3 57 12" src="https://github.com/next-step/java-lms/assets/23239084/a0efc496-5d09-4558-8cdf-f9fbb4bffb4b">

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,80 @@
+        if (freeOrPaid == FREE) {
+            return true;
+        }
+        return payment.amount() == sessionFee;
```
Payment 객체에 물어보도록 하면 어떨까요 :) ?

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public boolean isEnrollmentAmountValid(Payment payment) {
+        if (freeOrPaid == FREE) {
```
이부분도 FreeOrPait 객체에게 물어보면 좋을 것 같아요 :)

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Sessions.java:
```diff
@@ -0,0 +1,12 @@
+
+public class Sessions {
+
+    private final List<Session> sessions;
```
강의를 중복해서 넣을수도 있겠군요..! 👀

yoohojeong replied:
중복 생성을 막기 위해 List에서 Set으로 타입 변경했습니다~!

ohtaeg commented on src/main/java/nextstep/courses/domain/enrollment/Enrollment.java:
```diff
@@ -0,0 +1,38 @@
+        sessionUsers.addSessionUser(new SessionUser(nsUser, session));
+    }
+
+    public Enrollment(Session session) {
+        this(null, null, session);
```
해당 생성자로 생성 호출을 하게 된다면 null pointer 예외가 발생할 수 도 있을 것 같네요 :)

```java
    @ParameterizedTest
    @EnumSource(mode = Mode.EXCLUDE, names = {"PREPARING"})
    void 생성_테스트(SessionStatus sessionStatus) {
        Session session = new Session(sessionStatus);
        new Enrollment(session);
    }
```

<img width="438" alt="image" src="https://github.com/next-step/java-lms/assets/20432323/369a6d84-2cc8-4800-af1a-8a7ad63ef02e">

yoohojeong replied:
헙.. NPE,, 
해당 생성자를 지우고 아래 정적팩토리메소드로만 Enrollment를 생성할 수 있게 하였습니다~!

```java
public static Enrollment enroll(Payment payment, Session session, NsUser user) {
        return new Enrollment(payment, user, session);
    }
```

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,80 @@
+import nextstep.courses.domain.coverimage.CoverImage;
+import nextstep.payments.domain.Payment;
+
+public class Session {
```
현재 Session이 많은 필드를 가지고 있는데요! Session이 가지고 있는 책임을 줄여볼 수 있을까요 :) ?
다른 객체에게 필드를 일부 옮기거나 연관성 있는 필드들을 묶어 또 다른 객체를 도출해보시면 좋을 것 같아요!

yoohojeong replied:
`freeOrPaid:FreeOrPaid`와 `sessionFee:long`의 기능을 `sessionFee:SessionFee`로 합쳐서.. 필드를 조금 줄였습니다 ... !!!

ohtaeg replied:
의도한대로 잘해주셨네요 👍 👍 
추가로 무료강의와 유료강의에 따라 수강 인원도 달라지기 때문에 limitedEnrollment 필드도 같이 묶어서 관리하는 것도
괜찮을 것 같기도 하네요!! 이 부분은 꼭 수정해주실 필요는 없기에 한번 고민해보아요 :)

yoohojeong replied:
오 좋은 생각입니다. 안그래도 limitedEnrollment 위치가 애매해서 고민하다가 그냥 뒀는데..
바로 SessionFee로 이동 했습니다. ㅎㅎ

ohtaeg commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -0,0 +1,80 @@
+        this.sessionStatus = sessionStatus;
+    }
+
+    public Session(Period period) {
```
해당 생성자는 사용하지 않는 것 같아요 👀

ohtaeg reviewed (changes requested):
도메인들을 잘 만들어주셨네요 👍 👍 
소소한 코멘트들을 드렸는데 확인해주시면 감사하겠습니다!!
반영 후 재요청 부탁드립니다  😄

yoohojeong committed:
feat : Session UUID 제거

yoohojeong committed:
feat : #368 피드백 반영
- SessionUsers 중복 생성 막기 위해 List에서 Set으로 타입 변경
- Course, Session 생성 시, id 값 null 로 주입
- Session 내 일부 로직 Payment, FreeOrPaid로 이동
- Sessions 중복 생성 막기 위해 List에서 Set으로 타입 변경
- SessionStatus가 ENROLLMENT_OPEN 일 때, Enrollment(Session) NPE 이슈 해결
- 사용하지 않는 Session(Period) 생성자 제거

yoohojeong committed:
feat : #368 피드백 반영(2)
- Session 내 'freeOrPaid:FreeOrPaid'와 'sessionFee:long'의 기능을 'sessionFee:SessionFee'로 합침

yoohojeong committed:
style : Readme 변경

ohtaeg reviewed (approved):
리뷰가 늦어서 죄송합니다 🙏 
피드백 반영을 잘해주셨습니다 👍 👍 
얼마 남지 않았네요!! 다음 미션도 화이팅입니다 🔥
